src/lib/hooks/useEnableKeyboardController.tsx

```diff
@@ -0,0 +1,87 @@
+import {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react'
+import {KeyboardProvider} from 'react-native-keyboard-controller'
+import {useFocusEffect} from '@react-navigation/native'
+
+import {IS_DEV} from '#/env'
+
+const KeyboardControllerRefCountContext = createContext<{
+  incrementRefCount: () => void
+  decrementRefCount: () => void
+}>({
+  incrementRefCount: () => {},
+  decrementRefCount: () => {},
+})
+
+export function KeyboardControllerProvider({
+  children,
+}: {
+  children: React.ReactNode
+}) {
+  const [refCount, setRef] = useState(0)
+
+  const value = useMemo(
+    () => ({
+      incrementRefCount: () => setRef(count => count + 1),
+      decrementRefCount: () =>
+        setRef(count => {
+          const next = count - 1
+          if (IS_DEV && next < 0) {
+            console.error('KeyboardController ref count < 0')
+          }
+          return next
+        }),
+    }),
+    [],
+  )
+
+  return (
+    <KeyboardControllerRefCountContext.Provider value={value}>
+      <KeyboardProvider
+        enabled_PATCHED={refCount > 0}
```

Comment:
I slightly don't understand a full motivation of the patch 🤔 

In my head I'm confident that next solution should work:
- you just mount `KeyboardProvider` with default `enabled={false}` value (no patch needed here)
- you keep your code but instead of changing `enabled_PATCHED` property you are using `setEnabled()` from `useKeyboardController` and in `useEffect` call the same method? So the code will look like:

```tsx
const {setEnabled} = useKeyboardController();
// other code with reference count
useEffect(() => {
  setEnabled(refCount > 0);
}, [refCount])
```

Reply:
Setting state directly in effects is an anti-pattern because it causes an entire rendering cycle in vain. If it's possible to control a prop directly, it's always better.

Reply:
> Setting state in useEffect is messy

Well, if you think about internals of `setEnabled`, the yes, it's a react-antipattern (because internally `setEnabled` is based on `useState`). But I highly doubt it can produce infinte/loop state updates or something like that 🤷‍♂️ 

But if you just think about `useEffect` as a thing that runs effect if props gets changes, then I think it's okay to have such code? Because this is exactly what you want to achieve (to disable/enable module based on ref count)

Reply:
@gaearon okay, if you are concerned about double state updates, then can't you use `useRef` for ref counting and call `setEnabled` in `increment`/`decrement` methods? In this case you will not update state inside effect + you will have only a single state update? 🤔

Reply:
Yeah I think that would work. Just seems unfortunate that this is almost 1:1 mirror of a controlled component pattern (like input takes controlled value and uncontrolled initialValue)

Reply:
Will do 👍

Reply:
> Yeah I think that would work. Just seems unfortunate that this is almost 1:1 mirror of a controlled component pattern (like input takes controlled value and uncontrolled initialValue)

Well, I understand your frustration here... But `setEnabled` functionality was added only to use in some rare cases. From what I remember you started to disable `keyboard-controller` in https://github.com/bluesky-social/social-app/pull/4399 and then started to disable it no more and more screens.

I don't know why you disabled lib in other cases, but the original problem https://github.com/bluesky-social/social-app/pull/4399 from has been fixed in latest versions of keyboard-controller (originally fixed in 1.12.0 and a better fix that handles even more corner cases landed in 1.14.1)

So if it was the only one issue (incorrectly working keyboard-controller in Modals on Android), then this problem should be fixed in the lib itself.

If you have more use-cases where enabled keyboard-controller caused more issues and you had to turn it on and off, then let me know - will be happy to fix it in the package 😊 

Since your codebase is open it's very easy for me to reproduce the problem on my end 👀

Reply:
Fair enough, we can give it another try soon!

Reply:
Yeah, I have another branch where I enable it globally. Ran into some issues but nothing major!